source/clear-linux/tutorials/kubernetes.rst

```diff
@@ -0,0 +1,303 @@
+
+   .. code-block:: bash
+
+      sudo systemctl restart systemd-sysctl
+
+#. Enable kubelet service:
+
+   .. code-block:: bash
+
+      sudo systemctl enable kubelet.service
+
+#. Disable swap using one of the following methods, either:
+
+   a) Temporarily:
+
+      .. code-block:: bash
+
+         sudo swapoff -a
+
+      .. note::
+
+         Swap will be enabled at next reboot, causing failures in
+         your cluster.
+
+   or:
+
+   b) Permanently:
+
+      Mask the swap partition:
+
+      .. code-block:: bash
+
+         sudo systemctl mask `sed -n -e 's#^/dev/\([0-9a-z]*\).*#dev-\1.swap#p' /proc/swaps
+         sudo swapoff -a
+
+      .. note::
+
+         On systems with limited resources, some performance degradation may
+         be observed while swap is disabled.
+
+#. Configure the Kubernetes runtime interface, either:
+
+   a) Run Kubernetes with Docker + runc:
+
+      #. Enable docker.service:
+
+         .. code-block:: bash
+
+            sudo systemctl enable docker.service
+
+      #. Create (or edit if it exists) the file :file:`/etc/systemd/system/docker.service.d/51-runtime.conf` and include the following lines:
+
+         .. code-block:: bash
+
+            [Service]
+            Environment="DOCKER_DEFAULT_RUNTIME=--default-runtime runc"
+
+      #. Create (or edit if it exists) the file :file:`/etc/systemd/system/kubelet.service.d/10-kubeadm.conf` and include the following lines:
+
+         .. code-block:: bash
+
+            [Service]
+            Environment="KUBELET_EXTRA_ARGS="
+
+      #. Restart services:
+
+         .. code-block:: bash
+
+            sudo systemctl daemon-reload
+            sudo systemctl restart docker
+            sudo systemctl restart kubelet
+
+   or:
+
+   b) Run Kubernetes with CRI-O + your desired runtime. You can use multiple
+      runtimes with CRI-O, including *runc* and *kata-runtime*. To use
+      CRI-O + *kata-runtime*:
+
+      #. Enable crio.service:
+
+         .. code-block:: bash
+
+            sudo systemctl enable crio.service
+
+      #. Restart services:
+
+         .. code-block:: bash
+
+            sudo systemctl restart crio
+            sudo systemctl restart kubelet
+
+Run Kubernetes for the first time
+*********************************
+
+#. Prepare your system to run Kubernetes for the first time with the
+   following commands:
+
+   .. code-block:: bash
+
+      sudo systemctl daemon-reload
+      sudo systemctl restart crio
+      sudo systemctl restart docker
+
+#. Initialize the master control plane with the following command, either:
+
+   a) If you are running Kubernetes with Docker + runc:
+
+      .. code-block:: bash
+
+         sudo -E kubeadm init --pod-network-cidr 10.244.0.0/16 --ignore-preflight-errors=SystemVerification
+
+   or:
+
+   b) If you are running Kubernetes with CRI-O + kata-runtime:
+
+      .. code-block:: bash
+
+         sudo -E kubeadm init --pod-network-cidr 10.244.0.0/16 --cri-socket=/run/crio/crio.sock
+
```

Comment:
This seems like a regression due to the latest kata-runtime update to 1.3.0. because the runtime_untrusted_workload is set to kata-runtime during kata packages installation. I'll check this first thing tomorrow.
The kubelet defaults are already set in /usr/lib/systemd/system/kubelet.service. There was a debate on including detailed instructions to create override files but we decided to just include the "Package configuration customization in Clear Linux " section with a link to the Stateless man pages.

Reply:
I've verified that in CL version 25380 (and up), when installing containers-virt bundle, /etc/crio/crio.conf file is created if no previous one exists. This file contains the runtime_untrusted_workload setting to kata-runtime.

Reply:
Strange.  My file got updated and now has that set but my kata-runtime is still at 1.2. My clear linux is at  25400 right now.

Reply:
@eadamsintel: Yes, kata-runtime got back to 1.2 due to some pending issues with version 1.3. The fixes are WIP, but for this tutorial purposes, I think we are good now :)

Reply:
My clear linux is at 25400 too. 
When I view: /usr/share/defaults/crio/crio.conf, I still see: 
`runtime_untrusted_workload = "" `

@joselamego What is the consensus opinion here?

Reply:
There's a bug :(

First, virt-containers bundle adds ```/etc/crio/crio.conf``` but the new kata-runtime won't update the file if it already exists. Therefore ```runtime_untrusted_workload = "/usr/bin/kata-runtime"``` setting remains no matter the OS version. 

Removing  ```/etc/crio/crio.conf``` eventually gives what is needed.

Reply:
That is intentional. On a fresh installation there won't be a conf file in `/etc/crio/` so the `kata-runtime` is set. If user creates a customized file, then any update won't overwrite the user configuration. In case of need/desire, the `/etc/crio/` can be safely deleted and the default configuration goes in (stateless goals).

Reply:
On a second thought, you may be talking about the case of updating from pre-25380 to 25400-ish, so the wrong `/etc/crio/crio.conf` is there. In that case, yes, either manually deleting the previous file or uninstalling the bundle is required. Thoughts?

Reply:
I started with 25380 and swear that the crio.conf was missing the untrusted workload pointing to kata-runtime.  When my system updated the crio.conf got updated to and it was in /etc/crio   The time stamps of /etc/crio and my crio.conf match and are over an hour newer than two packages (rpm, lshw) that was updated with 25400.  There might be another issue with /etc files getting updated.   I might try to set this up again just to confirm.

Reply:
@eadamsintel Yes, please do so, and let us know the results. 
FYI: I've been running CL 25400 for quite some time. I have `containers-virt` installed, but I have not had to update. Therefore, it would be nice to investigate the possibility of whether /etc files are being updated or not in the update process.

Reply:
OK I installed 25380 and then after it got upgraded to 25410 I forced a downgrade to 25380 and disabled update before I installed any bundles.  Once I get to the step to restart the crio service it creates the crio.conf file and it has the untrusted workload pointing to the kata-runtime.   I don't know why it was not there before but it seems this is a non issue for now.

Reply:
Also, I upgrade to 25410 and have kata-runtime v1.3. The control plane still does not initialize to me.  I checked journalctl and I see errors about the kubelet.service failing.  The first error of this after it said the restart counter is at 6..  kuberlet.service: Start request repeated too quickly.

Reply:
@eadamsintel: There is a bug: Devicemapper support (among others) was added to crio, so now there are several storage options but no default defined in crio.conf. We are developing a patch to fix this. Should be available in tomorrow's release. 
Workaround for now is to set in `/etc/crio/crio.conf` the default storage driver: `storage_driver = "overlay"`
I've done this and my control plane comes up. Could you test, please?

Reply:
@joselamego I just tested the cluster setup with a fresh board and no container images on the disk and saw no errors. What makes you think that's the problem?

Reply:
Also,

I realized we've been talking about wrong setting.  My bad.

Setting ```runtime_untrusted_workload = "/usr/bin/kata-runtime"``` is right.

```default_workload_trust```should be ```trusted```.  You can check CRI-O's settings using ```crio config```.

@eadamsintel ping me offline/IRC and I'll help you to resolve the setup problem(s).

Reply:
@mythi: That error was preventing crio.service from starting at all. If your's did start correctly then either you didn't had the affected version or you've already get the fix. Either way I think we're good :)

Reply:
@joselamego I don't have the fix and my crio starts just fine.

Reply:
@eadamsintel I'd like set up a brief 1:1.  I want to know what, if any, details we should capture in the documentation.

Reply:
Clear ships `/usr/share/defaults/etc/hosts`

Reply:
True. It's not enough for kubelet that expects ```/etc/hosts```

Reply:
I would add it to the main doc to update the hosts.  Would 127.0.0.1 universally work for everyone out of the box?  As a side note the troubleshooting would be more helpful if more of the error was listed.  Noticing <hostname> not found is hard to see with all the logs.  This is the full error message.  I see a lot of "connect: connection refused" too that might be useful to add to the troubleshooting section. 

Oct 05 19:35:58 kata8 kubelet[1036]: E1005 19:35:58.266843    1036 eviction_manager.go:243] eviction manager: failed to get get summary stats: failed to get node info: node "kata8" not found

Reply:
I'm doing a fresh install with `25430` and walking through tut one more time. I'm going to check: 
-  if the `crio-conf` created the ‘overlay.’ 
- If before Step 4, we should add a section including the commands to access the ‘hosts’ file.

Reply:
I installed CL 25430 and upgraded to 25470. 

After following the tutorial, `/usr/share/defaults/crio/crio.conf` showed:
`storage_driver - "overlay"`.  Confirmed. That should address @eadamsintel concerns. 

However, this was blank:  `runtime_untrusted_workload = ""`

Next, there were several issues with Step 2 of *Run Kubenetes for the first time*, 
(I encountered a similar error as @eadamsintel)  

Command: 
`sudo -E kubeadm init --pod-network-cidr 10.244.0.0/16 --cri-socket=/run/crio/crio.sock`

Error: 

: unable to get URL "https://dl.k8s.io/release/stable-1.11.txt": Get https://dl.k8s.io/release/stable-1.11.txt: dial tcp 23.236.58.218:443: i/o timeout

Could there be a problem with the URL or port of the host server in command, Step 2? 
Has anyone tested this outside the internal firewall? 

I tried to troubleshoot this by attempting to test the host and port.  I received these errors: 

*Testing the TCP in error output*

If I run: `curl -S 23.236.58.218: 443`
I receive the error: 
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.10.3</center>
</body>
</html>
curl: (7) Couldn't connect to server

*Testing the host and port using command Step 2 in tutorial*
If I run: 
`curl -S https://10.244.0.0:16`
`curl -S https://10.244.0.0:80`

I receive the error: 
`curl: (7) Failed to connect to 10.244.0.0 port 80: Connection timed out.`

See my next comment too, please.

Reply:
IMHO:  I think this tutorial would be much easier to follow if these two paths were treated as separate but continuous sections: 

- Run Kubernetes with Docker + runc:
- Run Kubernetes with CRI-O + kata-runtime

Unfortunately, it's possibile that a user could mistakenly mismatch the either/or choices given. 

See sections: 
https://github.com/clearlinux/clear-linux-documentation/blob/2807437b32d9f13d7e001fac108df84a3895f491/source/clear-linux/tutorials/kubernetes.rst#configure-kubernetes

https://github.com/clearlinux/clear-linux-documentation/blob/2807437b32d9f13d7e001fac108df84a3895f491/source/clear-linux/tutorials/kubernetes.rst#run-kubernetes-for-the-first-time

Reply:
> However, this was blank: runtime_untrusted_workload = ""

This is expected: `/usr/share/defaults/crio/crio.conf` remains as you decribed, but `/etc/crio/crio.conf` is created when either _containers-virt_ or _cloud-native-basic_ are installed and it contains the appropriate setting of `runtime_untrusted_workload =  "/usr/bin/kata-runtime"`

Reply:
Regarding the URL issues, sound like proxies. What do you get if run `sudo -E kubeadm config images pull` ?

Reply:
> IMHO: I think this tutorial would be much easier to follow if ...

I agree. It should be a matter or reorganizing the existing paragraphs only. @mvincerx: would you willing to handle this?

Reply:
@joselamego Yes, either I or @MCamp859 can handle reorganizing the document for better flow.

Reply:
@joselamego. In response to running a different command, I receive the same error as before: 

If I run:
`sudo -E kubeadm config images pull`
I receive the error: 
: unable to get URL "https://dl.k8s.io/release/stable-1.11.txt": Get https://dl.k8s.io/release/stable-1.11.txt: dial tcp 23.236.58.218:443: i/o timeout

For me, this still leaves my question, above, unanswered. Has anyone tried to run this outside the firewall (in order to eliminate the possibility that this is a proxy config issue)?

Reply:
@joselamego. Thanks for clarifying the difference re `runtime_untrusted_workload` assignment in this location: /etc/crio/crio.conf.

Reply:
@mvincerx: I've executed the command in a system outside the corporate network, but won't hurt to hear other's experience. That said, we should be able to configure your system: when running `echo $HTTPS_PROXY`, you should get the corporate's proxy and the port i.e.: `proxy-something.company.com:XXX`. That is what _kubeadm_ reads (in capitals). Feel free to ping me and we can try live debugging .

Reply:
@joselamego I will send you an invitation for a meeting soon. Need to SSH into my NUC to test; that requires a little setup.

Reply:
@joselamego By *executed the command*, do you mean the command was executed successfully?

Reply:
@mvincerx: yes, I have successfully run kubernetes clusters in Clear Linux both inside and outside corporate network. Let's talk more about it on the meeeting.